src/hotspot/cpu/s390/stubGenerator_s390.cpp

```diff
@@ -1822,6 +1858,559 @@ class StubGenerator: public StubCodeGenerator {
+    __ z_stmg(Z_ARG2, Z_ARG4, argsave_offset,    parmBlk);
+
+    assert(AES_ctrVec_len > 0, "sanity. We need a counter vector");
+    __ add2reg(counter, AES_parmBlk_align, parmBlk);       // counter array is located behind crypto key. Available range is disp12 only.
+    __ z_mvc(0, AES_ctrVal_len-1, counter, 0, ctr);        // move first copy of iv
+    for (int j = 1; j < AES_ctrVec_len; j+=j) {            // j (and amount of moved data) doubles with every iteration
+      int offset = j * AES_ctrVal_len;
+      if (offset <= 256) {
+        __ z_mvc(offset, offset-1, counter, 0, counter);   // move iv
+      } else {
+        for (int k = 0; k < offset; k += 256) {
+          __ z_mvc(offset+k, 255, counter, 0, counter);
+        }
+      }
+    }
+
+    Label noCarry, done;
+    __ z_lg(scratch, Address(ctr, 8));                     // get low-order DW of initial counter.
+    __ z_algfi(scratch, AES_ctrVec_len);                   // check if we will overflow during init.
+    __ z_brc(Assembler::bcondLogNoCarry, noCarry);         // No, 64-bit increment is sufficient.
+
+    for (int j = 1; j < AES_ctrVec_len; j++) {             // start with j = 1; no need to add 0 to the first counter value.
+      int offset = j * AES_ctrVal_len;
+      generate_increment128(counter, offset, j, scratch);  // increment iv by index value
+    }
+    __ z_bru(done);
+
+    __ bind(noCarry);
+    for (int j = 1; j < AES_ctrVec_len; j++) {             // start with j = 1; no need to add 0 to the first counter value.
+      int offset = j * AES_ctrVal_len;
+      generate_increment64(counter, offset, j);            // increment iv by index value
+    }
+
+    __ bind(done);
+
+    BLOCK_COMMENT("} prepare stack counterMode_AESCrypt");
+  }
+
+
+  void generate_counterMode_increment_ctrVector(Register parmBlk, Register counter, Register scratch, bool v0_only) {
+
+    BLOCK_COMMENT("increment ctrVector counterMode_AESCrypt {");
+
+    __ add2reg(counter, AES_parmBlk_align, parmBlk);       // ptr to counter array needs to be restored
+    for (int j = 0; j < AES_ctrVec_len; j++) {
+      int offset = j * AES_ctrVal_len;
+      generate_increment128(counter, offset, AES_ctrVec_len, scratch); // increment iv by # vector elements
+      if (v0_only) break;
+    }
+
+    BLOCK_COMMENT("} increment ctrVector counterMode_AESCrypt");
+  }
+
+  void generate_increment64(Register counter, int offset, int increment) {
+    __ z_algsi(offset + 8, counter, increment);            // increment, no overflow check
+  }
+
+  void generate_increment128(Register counter, int offset, int increment, Register scratch) {
+    __ clear_reg(scratch);                                 // prepare to add carry to high-order DW
+    __ z_algsi(offset + 8, counter, increment);            // increment low order DW
+    __ z_alcg(scratch, Address(counter, offset));          // add carry to high-order DW
+    __ z_stg(scratch, Address(counter, offset));           // store back
+  }
```

Comment:
Good. Maybe add a Big Endian comment?

Reply:
With Big-Endian becoming more and more exotic, I agree it might be a good idea to explain what it means. I added a three-line comment preceding the generate_increment*() emitters. Will become visible with the next commit.